File: ios/Classes/NativeMoPubController.swift
//
//  NativeAdmobController.swift
//  flutter_native_admob
//
//  Created by Dao Duy Duong on 3/14/20.
//

import UIKit
import MoPub
import MoPub_AdMob_Adapters
import MoPub_FacebookAudienceNetwork_Adapters

class NativeMoPubController: NSObject, MPNativeAdDelegate {
    
    enum CallMethod: String {
        case setAdUnitID
        case reloadAd
    }
    
    enum LoadState: String {
        case loading, loadError, loadCompleted
    }
    
    let id: String
    let channel: FlutterMethodChannel
    
    var nativeAdChanged: ((MPNativeAd?) -> Void)?
    var nativeAd: MPNativeAd?
    
    private var adRequest: MPNativeAdRequest?
    private var adUnitID: String?
    
    private var isLoading = false
    
    init(id: String, channel: FlutterMethodChannel) {
        self.id = id
        self.channel = channel
        super.init()
        
        channel.setMethodCallHandler(handle)
    }
    
    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let callMethod = CallMethod(rawValue: call.method) else { return result(FlutterMethodNotImplemented) }
        let params = call.arguments as? [String: Any]
        
        switch callMethod {
        case .setAdUnitID:
//            print("[MoPub] ========= set ad and load ad =========")
            
            guard let adUnitID = params?["adUnitID"] as? String else {
                return result(nil)
            }
            
            var postCode: Float?
            var postCity: Int?
            if let postcode = params?["postCode"] as? Float {
                postCode = postcode
            }
            
            if let postcity = params?["postCity"] as? Int {
                postCity = postcity
            }
            
            let isChanged = adUnitID != self.adUnitID
            self.adUnitID = adUnitID
            
            if nativeAd == nil || isChanged {
                guard !isLoading else {
//                    print("[MoPub] isLoading \(self.adUnitID)")
                    channel.invokeMethod(LoadState.loading.rawValue, arguments: nil)
                    return
                }
                isLoading = true
                let settings = MPStaticNativeAdRendererSettings.init()
                settings.renderingViewClass = MopubNativeAdView.self
                let config = MPStaticNativeAdRenderer.rendererConfiguration(with: settings)
                let googleConfig = MPGoogleAdMobNativeRenderer.rendererConfiguration(with: settings)
                let facebookConfig = FacebookNativeAdRenderer.rendererConfiguration(with: settings)
                adRequest = MPNativeAdRequest.init(adUnitIdentifier: adUnitID, rendererConfigurations: [config, googleConfig, facebookConfig])
                
//                print("[MoPub] loadAd: \(self.adUnitID), \(self.nativeAd)")
                loadAd(postCode, postCity)
            } else {
//                print("[MoPub] has ad \(self.adUnitID)")
                invokeLoadCompleted()
            }
            
        case .reloadAd:
            var postCode: Float?
            var postCity: Int?
            if let postcode = params?["postCode"] as? Float {
                postCode = postcode
            }
            if let postcity = params?["postCity"] as? Int {
                postCity = postcity
            }
            
//            print("[MoPub] ========= reload ad =========")
            
            guard !isLoading else {
//                print("[MoPub] isLoading \(self.adUnitID)")
                channel.invokeMethod(LoadState.loading.rawValue, arguments: nil)
                return
            }
            
            nativeAd = nil
            isLoading = true
//            print("[MoPub] ad reload \(self.adUnitID)")
            if adRequest == nil {
                let settings = MPStaticNativeAdRendererSettings.init()
                settings.renderingViewClass = MopubNativeAdView.self
                let config = MPStaticNativeAdRenderer.rendererConfiguration(with: settings)
                let googleConfig = MPGoogleAdMobNativeRenderer.rendererConfiguration(with: settings)
                let facebookConfig = FacebookNativeAdRenderer.rendererConfiguration(with: settings)
                adRequest = MPNativeAdRequest.init(adUnitIdentifier: adUnitID, rendererConfigurations: [config, googleConfig, facebookConfig])
//                print("[MoPub]MPNativeAdRequest: \(self.adUnitID), \(self.adRequest)")
            }
            loadAd(postCode, postCity)
        }
        
        result(nil)
    }
    
    private func loadAd(_ postCode: Float?, _ postCity: Int?) {
//        print("[MoPub] ad loading \(self.adUnitID)")
        channel.invokeMethod(LoadState.loading.rawValue, arguments: nil)
        
        var data = ""
        
        if let postCode = postCode {
            data = "w_postCode:\(String(describing: postCode).replacingOccurrences(of: ".0", with: ""))"
        }
        
        if let postCity = postCity {
            if data == "" {
                data = "w_postCity:\(postCity)"
            } else {
                data = "\(data),w_postCity:\(postCity)"
            }
        }
        
        if data != "" {
            let targeting = MPNativeAdRequestTargeting.init()
            targeting?.userDataKeywords = data
            
            if let targeting = targeting {
                adRequest?.targeting = targeting
            }
        }
        
        adRequest?.start(completionHandler: { request, response, error in
            self.isLoading = false
            
            guard error == nil else {
//                print("[MoPub] NativeMoPub: \(self.adUnitID) failed to load with error: \(error)")
                self.channel.invokeMethod(LoadState.loadError.rawValue, arguments: nil)
                return
            }
            
            self.nativeAd = response
            self.nativeAd?.delegate = self
            self.invokeLoadCompleted()
            
            print("[MoPub] load ad complete \(self.adUnitID)")
        })
    }
    
    private func invokeLoadCompleted() {
        nativeAdChanged?(nativeAd)
        channel.invokeMethod(LoadState.loadCompleted.rawValue, arguments: nil)
    }
    
    func viewControllerForPresentingModalView() -> UIViewController! {
        let app = UIApplication.shared.delegate as! FlutterAppDelegate
        if let rootViewController = app.window.rootViewController{
            return rootViewController
        }else {
            return UIViewController()
        }
    }
    
}

class NativeMoPubControllerManager {
    
    static let shared = NativeMoPubControllerManager()
    
    private var controllers: [NativeMoPubController] = []
    
    private init() {}
    
    func createController(forID id: String, binaryMessenger: FlutterBinaryMessenger) {
        if getController(forID: id) == nil {
            let methodChannel = FlutterMethodChannel(name: id, binaryMessenger: binaryMessenger)
            let controller = NativeMoPubController(id: id, channel: methodChannel)
            controllers.append(controller)
        }
    }
    
    func getController(forID id: String) -> NativeMoPubController? {
        return controllers.first(where: { $0.id == id })
    }
    
    func removeController(forID id: String) {
        if let index = controllers.firstIndex(where: { $0.id == id }) {
            controllers.remove(at: index)
        }
    }
}

File: android/src/main/kotlin/com/nover/flutternativeadmob/NativeMopubController.kt
package com.nover.flutternativeadmob

import android.content.Context
import android.os.AsyncTask
import android.util.Log
import android.view.View
import com.facebook.ads.AdSettings
import com.google.android.gms.ads.formats.NativeAdOptions
import com.google.android.gms.ads.identifier.AdvertisingIdClient
import com.google.android.gms.common.GooglePlayServicesNotAvailableException
import com.google.android.gms.common.GooglePlayServicesRepairableException
import com.mopub.nativeads.*
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.IOException


class NativeMopubController(
        val id: String,
        private val channel: MethodChannel,
        private val context: Context
) : MethodChannel.MethodCallHandler, MoPubNative.MoPubNativeNetworkListener {

    enum class CallMethod {
        setAdUnitID, reloadAd
    }

    enum class LoadState {
        loading, loadError, loadCompleted
    }

    var nativeAdChanged: ((MoPubNative?) -> Unit)? = null
    var nativeAd: NativeAd? = null
        set(value) {
            field = value
            invokeLoadCompleted()
        }
    var adLoader: MoPubNative? = null

    private var adUnitID: String? = null

    init {
        channel.setMethodCallHandler(this)
//        val task: AsyncTask<Void?, Void?, String?> = object : AsyncTask<Void?, Void?, String?>() {
//            override fun onPostExecute(advertId: String?) {
//                Log.e("======================", advertId)
//            }
//
//            override fun doInBackground(vararg p0: Void?): String? {
//                var idInfo: AdvertisingIdClient.Info? = null
//                try {
//                    idInfo = AdvertisingIdClient.getAdvertisingIdInfo(context)
//                } catch (e: GooglePlayServicesNotAvailableException) {
//                    e.printStackTrace()
//                } catch (e: GooglePlayServicesRepairableException) {
//                    e.printStackTrace()
//                } catch (e: IOException) {
//                    e.printStackTrace()
//                }
//                var advertId: String? = null
//                try {
//                    advertId = idInfo!!.id
//                } catch (e: NullPointerException) {
//                    e.printStackTrace()
//                }
//                return advertId
//            }
//        }
//        task.execute()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (CallMethod.valueOf(call.method)) {
            CallMethod.setAdUnitID -> {
                call.argument<String>("adUnitID")?.let {
                    val isChanged = adUnitID != it
                    adUnitID = it

                    if (adLoader == null || isChanged) {
                        adLoader = MoPubNative(this.context, it, this)
                        val viewBinder: ViewBinder = ViewBinder.Builder(R.layout.mopub_native_ad_layout)
                                .mainImageId(R.id.ad_media)
                                .titleId(R.id.ad_headline)
                                .privacyInformationIconImageId(R.id.privacy_information_icon)
                                .callToActionId(R.id.ad_call_to_action)
                                .build()
                        val moPubStaticNativeAdRenderer = MoPubStaticNativeAdRenderer(viewBinder)

                        val googleViewBinder =
                                GooglePlayServicesViewBinder.Builder(R.layout.mopub_native_ad_layout_google)
                                        .mediaLayoutId(R.id.ad_media_google)
                                        .titleId(R.id.ad_headline)
                                        .privacyInformationIconImageId(R.id.privacy_information_icon)
                                        .callToActionId(R.id.ad_call_to_action)
                                        .addExtra(GooglePlayServicesAdRenderer.VIEW_BINDER_KEY_AD_CHOICES_ICON_CONTAINER,
                                                R.id.native_ad_choices_icon_container)
                                        .build()
                        val googlePlayServicesAdRenderer = GooglePlayServicesAdRenderer(googleViewBinder)

                        val facebookViewBinder =
                                FacebookAdRenderer.FacebookViewBinder.Builder(R.layout.mopub_native_ad_layout_facebook)
                                        .titleId(R.id.ad_headline)
                                        .mediaViewId(R.id.ad_media_facebook)
                                        .adChoicesRelativeLayoutId(R.id.privacy_information_icon)
                                        .advertiserNameId(R.id.ad_headline)
                                        .callToActionId(R.id.ad_call_to_action)
                                        .build()
                        val facebookAdRenderer = FacebookAdRenderer(facebookViewBinder)

                        adLoader?.registerAdRenderer(googlePlayServicesAdRenderer)
                        adLoader?.registerAdRenderer(facebookAdRenderer)
                        adLoader?.registerAdRenderer(moPubStaticNativeAdRenderer)
                    }

                    val postCode: Number? = call.argument<Number>("postCode")
                    val postCity: Int? = call.argument<Int>("postCity")

                    if (nativeAd == null || isChanged) loadAd(postCode, postCity) else invokeLoadCompleted()
                } ?: result.success(null)
            }

            CallMethod.reloadAd -> {
                call.argument<Boolean>("forceRefresh")?.let {
                    val postCode: Number? = call.argument<Number>("postCode")
                    val postCity: Int? = call.argument<Int>("postCity")
                    if (it || adLoader == null) loadAd(postCode, postCity) else invokeLoadCompleted()
                }
            }
        }
    }

    private fun loadAd(postCode: Number?, postCity: Int?) {
        channel.invokeMethod(LoadState.loading.toString(), null)
        // Facebook test device id
        AdSettings.addTestDevice("54c5265c-4eb1-4eb0-a74e-c146ebbcc282");

        // admob ad choice position
        val extras: HashMap<String, Any> = HashMap()
        extras[GooglePlayServicesNative.KEY_EXTRA_AD_CHOICES_PLACEMENT] = NativeAdOptions.ADCHOICES_TOP_RIGHT
        adLoader?.setLocalExtras(extras)

        var data = ""
        postCode?.let {
            data = "w_postCode:${it.toString().replace(".0", "")}"
        }
        postCity?.let {
            if(data == "") data = "w_postCity:${it}"
            else data = "${data},w_postCity:${it}"
        }

        if(data != "") {
            val parameters: RequestParameters =
                    RequestParameters.Builder()
                            .userDataKeywords(data)
                            .build()

            adLoader?.makeRequest(parameters)
        }
        else {
            adLoader?.makeRequest()
        }
    }

    private fun invokeLoadCompleted() {
        nativeAdChanged?.let { it(adLoader) }
        channel.invokeMethod(LoadState.loadCompleted.toString(), null)
    }

    override fun onNativeLoad(ad: NativeAd?) {
        Log.d("MoPub", "Native ad has loaded.")
        nativeAd = ad
        nativeAd?.setMoPubNativeEventListener(
                object : NativeAd.MoPubNativeEventListener {
                    override fun onClick(view: View?) {
                        Log.d("MoPub", "Native ad has clicked.")
                    }

                    override fun onImpression(view: View?) {
                        Log.d("MoPub", "Native ad has impressed.")
                    }

                })
    }

    override fun onNativeFail(errorCode: NativeErrorCode?) {
        Log.d("MoPub", "Native ad failed to load with error: " + errorCode.toString())
        channel.invokeMethod(LoadState.loadError.toString(), null)
    }
}

object NativeMopubControllerManager {
    private val controllers: ArrayList<NativeMopubController> = arrayListOf()

    fun createController(id: String, binaryMessenger: BinaryMessenger, context: Context) {
        if (getController(id) == null) {
            val methodChannel = MethodChannel(binaryMessenger, id)
            val controller = NativeMopubController(id, methodChannel, context)
            controllers.add(controller)
        }
    }

    fun getController(id: String): NativeMopubController? {
        return controllers.firstOrNull { it.id == id }
    }

    fun removeController(id: String) {
        val index = controllers.indexOfFirst { it.id == id }
        if (index >= 0) controllers.removeAt(index)
    }
}